interface/new/new_comprehensive.php

```diff
@@ -380,61 +414,73 @@ function validate(f) {
           return false;
         }
         <?php } ?>
-        var msg = "";
         var ins_error = '0';
-        msg += "<?php echo htmlspecialchars(xl('The following fields are required'),ENT_QUOTES); ?>:\n\n";
         for ( var i = 0; i < errMsgs.length; i++ ) {
-               msg += errMsgs[i] + "\n";
+          iziToast.error({
+            title: 'Error',
+            message: "<?php echo htmlspecialchars(xl('The following field is required: '),ENT_QUOTES); ?>" + errMsgs[i],
```

Comment:
xl or xlt ?

Reply:
@aethelwulffe i have also been wondering on this from time to time i know the `xl` , `xlt` are used for translation. What is the difference between the two

Reply:
xl() is defined in library/translations.inc.php, and has a number of functions that we don't actually use, like special label versions etc...
These functions are supposedly not compatible with HTMLSPECIALCHAR.
```
C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\interface\language\lang.info.html:
   62  IV. Developer Instructions
   63  
   64: The translation is done by the xl() custom function, which can be found
   65  at libreehr/library/translation.inc.php.
   66  Its semantics are very simple.
   ..
   85  http://librehealth.io/wiki/index.php/Development_Policies#Internationalization
   86  
   87: There are also wrappers to the xl() function that are used infrequently in specific situations. Please
   88  refer to libreehr/library/translation.inc.php for more details.
   89  

C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\interface\orders\orders_results.php:
   87      list($order_id, $order_seq, $report_id, $result_id) = explode(':', $line_value);
   88  
   89:     // Not using xl() here because this is for debugging only.
   90      if (empty($order_id)) die("Order ID is missing from line $lino.");
   91  

C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\library\htmlspecialchars.inc.php:
   93  	} else {
   94  		trigger_error(
   95: 			'Translation via xl() was requested, but the xl()'
   96  			. ' function is not defined, yet.',
   97  			E_USER_WARNING
   ..
  102  
  103  /**
  104:  * Translate via xl() and then escape via text().
  105   *
  106   * @param string $key The string to escape, possibly including "&", "<",
  ...
  113  
  114  /**
  115:  * Translate via xl() and then escape via attr().
  116   *
  117   * @param string $key The string to escape, possibly including (&), (<),
  ...
  124  
  125  /*
  126: Translate via xl() and then escape via addslashes for use with javascript literals
  127  */
  128  function xls($key){

C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\library\translation.inc.php:
   53  }
   54  
   55: // ----------- xl() function wrappers ------------------------------
   56  //
   57: // Use above xl() function the majority of time for translations. The
   58  //  below wrappers are only for specific situations in order to support
   59  //  granular control of translations in certain parts of LibreEHR.

C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\library\authentication\privDB.php:
   81      if ($recordset === FALSE) {
   82          
   83:       // These error messages are explictly NOT run through xl() because we still
   84        // need them if there is a database problem.
   85        echo "Failure during database access! Check server error log.";

C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\library\Smarty\plugins\function.xl.php:
    4   * @package Smarty
    5   * @subpackage plugins
    6:  * simplified xl() version for smarty templates
    7   * Christian Navalici 2007
    8   */

C:\AlphaFiles\bidenis\REPOS\BRANCH_TESTS\modules\language_translations\collectConstants.pl:
  208     }
  209  
  210:    # skip $. Raally rare usage of xl() function.
  211     # There are about 25 lines of this in entire codebase
  212     # and likely just several contants. Can put in manually
  ...
  230     # Skip line with the js_xl function
  231     if ($de eq "m") {
  232:     print LOGFILE "MESSAGE:  Special case character 'm' skipped. Done to skip the function js_xl() line.\n";
  233      print LOGFILE $editvar2."\n";
  234      next;
```
You can find the xlt() in htmlspecialchars.inc.php here:
```
C:\AlphaFiles\bidenis\REPOS\MIPS_Gateway\library\htmlspecialchars.inc.php:
  108   * @return string The string, with "&", "<", and ">" escaped.
  109   */
  110: function xlt($key) {
  111  	return text(hsc_private_xl_or_warn($key));
  112  }
```

Reply:
so which are we to use henceforth?

Reply:
The functions in htmlspecialchars.inc.php (xlt, xla etc...) include special character functions sorta.  Use what you need, just know what those are.
xlt() includes a function also defined in htmlspecialchars.inc.php that handles that stuff, and is not supposed to be called directly (hidden).